<launch>

  <arg name="worldfile" default="$(find exploration_ros)/stagefiles/hospital.world"/>
  <arg name="action" default="explore"/>
  <arg name="robotname" default="my_robot"/>
  <arg name="laser_topic" default="scan"/>
  <arg name="laser_frame" default="base_laser_link"/>
  <arg name="base_frame" default="base_link"/>
  <master auto="start"/>
  <param name="/use_sim_time" value="true"/>

  <node pkg="exploration_ros" name="slam" type="slam_node" output="screen">

    <remap from="scan" to="$(arg robotname)/$(arg laser_topic)" />

    <param name="resolution" value="0.025" /> <!-- doc="resolution of the matching grid"  -->
    <param name="maxScore" value="0.15" /> <!-- doc="score of the matcher, the higher the less matches" /> -->
    <param name="kernelRadius" value="0.2" /> <!-- doc="radius of the convolution kernel" /> -->
    <param name="minInliers" value="5" /><!--  doc="min inliers"/> -->
    <param name="windowLoopClosure" value="10" /> <!-- doc="sliding window for loop closures"  /> -->
    <param name="inlierThreshold" value="2" /> <!-- doc="inlier threshold" /> -->
    <param name="angularUpdate" value="0.5" /> <!-- doc="angular rotation interval for updating the graph, in radians" default="0.7853975"/> -->
    <param name="linearUpdate"  value="0.5" /> <!-- doc="linear translation interval for updating the graph, in meters" default="0.25" /> -->
    <param name="timeUpdate"  value="0" /> <!-- doc="interval in time for updating the robot pose in the map, in seconds" default="0" /> -->
    <param name="type" value="0" /> <!-- doc="0 if stage simulation or 1 if real robot experiment"  default="0"  /> -->
    <param name="laserFrame" value="base_laser_link" /> <!-- doc="name of the laser TF frame"/> -->
    <param name="odometryFrame" value="$(arg robotname)/odom" /> <!-- doc="odometry ROS frame"/> -->
    <param name="scanTopic" value="$(arg robotname)/$(arg laser_topic)" /> <!-- value="scan" doc="scan ROS topic"/> -->
    <param name="occupancyTopic" value="/map" /> <!-- doc="occupancy grid ROS topic"/> -->

  </node>


    <!--
    <include file="$(find exploration_ros)/move_base_config/slam_gmapping.xml"/>
  -->

  <node pkg="exploration_ros" name="explorer_server" type="exploration_action_node" output="screen">  
    <param name="action" value="$(arg action)" />
    <param name="mapFrame" value="map" /> <!--  doc="TF mapFrame for the robot" --> 
    <param name="baseFrame" value="base_link" />  <!-- doc="TF base Frame for the robot" -->
    <param name="laserFrame" value="base_laser_link" /> <!-- doc="TF laser Frame for the robot" -->
    <param name="scanTopic" value="scan" /> <!--  doc="laser scan ROS topic" -->
    <param name="pointsTopic" value="points" /> <!-- doc="frontier points ROS topic" -->
    <param name="markersTopic" value="markers" /> <!-- doc="frontier centroids ROS topic" -->
    <param name="regionSize" value="15" /> <!-- doc="minimum size of a frontier region" -->
    <param name="exploredArea" value="10" /> <!-- doc="minimum number of visible frontier points before aborting a goal" -->
    <param name="lambda" value="1.25" /> <!-- doc="distance decay factor for choosing next goal" -->
    <param name="mc" value="0.5" /> <!-- doc="Lower distance limit to consider 2 goals as the same, in meters" -->
    <param name="Mc" value="5.0" /> <!-- doc="Max distance at which a centroid is considered if there are also closer ones, in meters"  -->
    <param name="nc" value="10" /> <!-- doc="Maximum number of centroids considered during each search for a goal" -->
    <param name="iter" value="-1" /> <!-- doc="Number of plans to be computed. -1 means infinite" -->
  </node> 

  <node pkg="exploration_ros" name="explorer_client" type="exploration_action_client" output="screen">
    <param name="action" value="$(arg action)" />
  </node> 

  <node pkg="stage_ros" type="stageros" name="stageros" args="$(arg worldfile)" respawn="false" >
    <param name="base_watchdog_timeout" value="0.2"/>
  </node>


  <node pkg="rviz" type="rviz" name="visualizer" output="screen" args="-d $(find exploration_ros)/config.rviz"/>


  <!-- <include file="$(find exploration_ros)/move_base_config/move_base.xml"/> -->
  <node pkg="move_base" type="move_base" respawn="false" name="move_base_node" output="screen">
    <remap from="map" to="/map" />
    <remap from="odom" to="odom" />
    <remap from="cmd_vel" to="cmd_vel" />

    <rosparam file="$(find exploration_ros)/move_base_config/move_base_params.yaml" command="load" />
    <rosparam file="$(find exploration_ros)/move_base_config/costmap_common_params.yaml" command="load" ns="global_costmap" />
    <rosparam file="$(find exploration_ros)/move_base_config/costmap_common_params.yaml" command="load" ns="local_costmap" />
    <rosparam file="$(find exploration_ros)/move_base_config/local_costmap_params.yaml" command="load" />
    <rosparam file="$(find exploration_ros)/move_base_config/global_costmap_params.yaml" command="load" />
    <rosparam file="$(find exploration_ros)/move_base_config/base_local_planner_params.yaml" command="load" />
    <rosparam file="$(find exploration_ros)/move_base_config/base_global_planner_params.yaml" command="load" />
    <rosparam file="$(find exploration_ros)/move_base_config/dwa_local_planner_params.yaml" command="load" />
    
    <!-- Override MOVE_BASE Frame Params to include prefix -->
    <param name="global_costmap/laser_scan_sensor/sensor_frame" value="/$(arg robotname)/$(arg laser_frame)"/>
    <param name="global_costmap/laser_scan_sensor/topic" value="/$(arg robotname)/$(arg laser_topic)"/>
    <param name="global_costmap/robot_base_frame" value="/$(arg robotname)/$(arg base_frame)"/>   
    <param name="local_costmap/global_frame" value="/$(arg robotname)/odom"/>
    <param name="local_costmap/laser_scan_sensor/sensor_frame" value="/$(arg robotname)/$(arg laser_frame)"/>
    <param name="local_costmap/laser_scan_sensor/topic" value="/$(arg robotname)/$(arg laser_topic)"/>
    <param name="local_costmap/robot_base_frame" value="/$(arg robotname)/$(arg base_frame)"/> 
  </node>


</launch>
